### PR TITLE
feat: add `emitLegacyCommonJSImports` flag

### PR DESCRIPTION
**Pull request by saihaj**
## Description

See changeset

Related #8067 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take action on it as appropriate

## How Has This Been Tested?

Updated `dev-tests`

## Checklist:

- [X] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

## Further comments

Couldn't figure out how to make it work for `oclif` plugin

### CODE REVIEW

**Review by n1ru4l — CHANGES_REQUESTED** [review]
Can you please resolve the failing tests (update the test fixtures)?

File: packages/plugins/typescript/gql-tag-operations/src/index.ts [hunk 1]
```diff
@@ -39,7 +40,7 @@ export const plugin: PluginFunction<{
       `\n`,
       ...getDocumentRegistryChunk(sourcesWithOperations),
```
[the remark this answers is not in the document]
Comment (charlypoly) [reply packages/plugins/typescript/gql-tag-operations/src/index.ts:41]: I'll fix it in upcoming commits

File: packages/plugins/other/visitor-plugin-common/src/imports.ts [hunk 2]
```diff
@@ -56,9 +57,14 @@ export function generateImportStatement(statement: ImportDeclaration): string {
     importSource.identifiers && importSource.identifiers.length
       ? `{ ${Array.from(new Set(importSource.identifiers)).join(', ')} }`
       : '*';
+  const importExtension =
+    importPath.startsWith('/') || importPath.startsWith('.') ? (statement.emitLegacyCommonJSImports ? '' : '.js') : '';
   const importAlias = importSource.namespace ? ` as ${importSource.namespace}` : '';
   const importStatement = typesImport ? 'import type' : 'import';
-  return `${importStatement} ${importNames}${importAlias} from '${importPath}';${importAlias ? '\n' : ''}`;
+  return `${importStatement} ${importNames}${importAlias} from '${importPath}${importExtension}';${
```
Comment (charlypoly) [inline packages/plugins/other/visitor-plugin-common/src/imports.ts:64]: @saihaj I saw that we are adding `.js` to `.d.ts` file imports, would it actually work?

ex: `import * as Types from '../types.d.js';` seems invalid to me

In short, do we need to add `.js` to `.d.ts` imports?

_/invoke @n1ru4l 👀_

File: packages/presets/near-operation-file/tests/near-operation-file.spec.ts [hunk 3]
```diff
@@ -6,6 +6,14 @@ import path from 'path';
 import { preset } from '../src/index.js';
 
 describe('near-operation-file preset', () => {
+  const executePreset: typeof preset.buildGeneratesSection = options =>
```
Comment (charlypoly) [inline packages/presets/near-operation-file/tests/near-operation-file.spec.ts:9]: fixed the behavior handled in `packages/graphql-codegen-cli/src/codegen.ts`

File: packages/graphql-codegen-cli/src/config.ts [hunk 4]
```diff
@@ -445,3 +445,19 @@ function addHashToDocumentFiles(documentFilesPromise: Promise<Types.DocumentFile
     })
   );
 }
+
+export function shouldEmitLegacyCommonJSImports(config: Types.Config, outputPath: string): boolean {
+  const globalValue = config.emitLegacyCommonJSImports === undefined ? true : !!config.emitLegacyCommonJSImports;
+  // const outputConfig = config.generates[outputPath];
+
+  // if (!outputConfig) {
```
Comment (charlypoly) [inline packages/graphql-codegen-cli/src/config.ts:453]: @saihaj we need to discuss whenever we want `emitLegacyCommonJSImports` to be also set at the GenerateConfig level 👀

File: packages/plugins/other/visitor-plugin-common/src/imports.ts (hunk 2 above)
Comment (n1ru4l) [reply packages/plugins/other/visitor-plugin-common/src/imports.ts:64]:
```
import * as Types from '../types.d.js';
```

is legit according to TypeScript module resolution `d.js` will point to `d.ts`. It is super confusing same as `.js` points to `.ts`, but that is how they decided upon.

Comment (charlypoly) [reply packages/plugins/other/visitor-plugin-common/src/imports.ts:64]: Okay, thanks @n1ru4l!

File: packages/graphql-codegen-cli/src/config.ts (hunk 4 above)
Comment (charlypoly) [reply packages/graphql-codegen-cli/src/config.ts:453]: Discussed together via Slack, let's keep it simple for now.

**Comment by charlypoly** [comment]
Close https://github.com/dotansimha/graphql-code-generator/pull/8094/files is this is merged first

**Review by n1ru4l — APPROVED** [review]
lfg_emoji.gif

**Pull request merged**